### PR TITLE
Remove stale egressip status entry

### PR DESCRIPTION
**Pull request by pperiyasamy**
This commit ensures egress IP status is in sync with cloud private IP config on a restarted ovnk master because stale entry is noticed in the egress IP status though cloud private IP config entry is not present for the given egress IP.  This might have happened while processing cloud private IP config delete event, an error is thrown because of bug fixed in PR: https://github.com/ovn-org/ovn-kubernetes/pull/3371.

### CODE REVIEW

**Comment by flavio-fernandes** [comment]
Reported-at: https://issues.redhat.com/browse/OCPBUGS-12747

File: go-controller/pkg/ovn/egressip.go
Comment (flavio-fernandes) [inline go-controller/pkg/ovn/egressip.go:1386]: nit: I know the other returns here are all doing `%v`, but let's at least do better with this one and use `%w`:
```
return fmt.Errorf("syncEgressIPs unable to sync egressip status: %w", err)
```

File: go-controller/pkg/ovn/egressip.go
Comment (flavio-fernandes) [inline go-controller/pkg/ovn/egressip.go:1630]: let's put an info long here about this update. Actually, it may be useful to have one in line 1626 too, wdyt?

File: go-controller/pkg/ovn/egressip.go
Comment (flavio-fernandes) [inline go-controller/pkg/ovn/egressip.go:1662]: add a comment here, explaining that cache must not include eips that cncc does not know about. Is this really needed? I see `oc.getCloudPrivateIPConfigMap()` being called in `syncEgressIPsStatus()` so I'm a bit confused here.

**Review by flavio-fernandes — COMMENTED** [review]
thanks for digging into this, @pperiyasamy ! 
Do you think we can add a unit test to exercise `syncEgressIPsStatus()` ?

**Comment by flavio-fernandes** [comment]
unit test failure details:
```
2023-05-18T13:01:25.8466867Z Hybrid Overlay Node Linux Operations [It] sets up local linux pod
2023-05-18T13:01:25.8467775Z /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/hybrid-overlay/pkg/controller/node_linux_test.go:515
2023-05-18T13:01:25.8468055Z
```

@JacobTanenbaum Is this the known flake?

```
...
2023-05-18T13:01:25.4833759Z
2023-05-18T13:01:25.4834351Z ^[[91m^[[1m• Failure [2.030 seconds]^[[0m
2023-05-18T13:01:25.4834789Z Hybrid Overlay Node Linux Operations
2023-05-18T13:01:25.4836145Z ^[[90m/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/hybrid-overlay/pkg/controller/node_linux_test.go:265^[[0m
2023-05-18T13:01:25.4836690Z   ^[[91m^[[1msets up local linux pod [It]^[[0m
2023-05-18T13:01:25.4837253Z   ^[[90m/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/testing/testing.go:18^[[0m
2023-05-18T13:01:25.4837773Z
2023-05-18T13:01:25.4837966Z   ^[[91mTimed out after 2.000s.
2023-05-18T13:01:25.4838282Z   Executed commands do not match expected commands!
2023-05-18T13:01:25.4838818Z   Executed unexpected command /fake-bin/ovs-ofctl dump-flows --no-stats br-ext table=20
2023-05-18T13:01:25.4839129Z
2023-05-18T13:01:25.4839565Z   [00]   /fake-bin/ovn-nbctl --timeout=15 get logical_switch mynode other-config:subnet
2023-05-18T13:01:25.4840297Z   [01]   /fake-bin/ovs-vsctl --timeout=15 --may-exist add-br br-ext -- set Bridge br-ext fail_mode=secure -- set Interface br-ext mtu_request=1400
2023-05-18T13:01:25.4840930Z   [02]   /fake-bin/ovs-vsctl --timeout=15 --if-exists get interface br-ext mac_in_use
2023-05-18T13:01:25.4841501Z   [03]   /fake-bin/ovs-vsctl --timeout=15 set bridge br-ext other-config:hwaddr=10:11:12:13:14:15
2023-05-18T13:01:25.4842488Z   [04]   /fake-bin/ovs-vsctl --timeout=15 --may-exist add-port br-int int -- --may-exist add-port br-ext ext -- set Interface int type=patch options:peer=ext external-ids:iface-id=int-mynode -- set Interface ext type=pa\
tch options:peer=int
2023-05-18T13:01:25.4843828Z   [05]   /fake-bin/ovs-vsctl --timeout=15 --may-exist add-port br-ext ext-vxlan -- set interface ext-vxlan type=vxlan options:remote_ip="flow" options:key="flow" options:dst_port=4789
2023-05-18T13:01:25.4844474Z   [06]   /fake-bin/ovs-ofctl dump-flows --no-stats br-ext table=20
2023-05-18T13:01:25.4844978Z   [07]   /fake-bin/ovs-ofctl -O OpenFlow13 --bundle replace-flows br-ext -
2023-05-18T13:01:25.4845459Z   [08]   /fake-bin/ovs-ofctl dump-flows --no-stats br-ext table=20
2023-05-18T13:01:25.4845942Z   [09]   /fake-bin/ovs-ofctl -O OpenFlow13 --bundle replace-flows br-ext -
2023-05-18T13:01:25.4846415Z   [10] ! /fake-bin/ovs-ofctl dump-flows --no-stats br-ext table=20
2023-05-18T13:01:25.4846899Z   [11] ! /fake-bin/ovs-ofctl -O OpenFlow13 --bundle replace-flows br-ext -
2023-05-18T13:01:25.4847183Z
2023-05-18T13:01:25.4847363Z   Expected
2023-05-18T13:01:25.4847618Z       <bool>: false
2023-05-18T13:01:25.4847881Z   to be true^[[0m
2023-05-18T13:01:25.4847994Z
2023-05-18T13:01:25.4848479Z   /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/hybrid-overlay/pkg/controller/node_linux_test.go:515
2023-05-18T13:01:25.4848934Z ^[[90m------------------------------^[[0m
```

**Comment by tssurya** [comment]
flake because of https://github.com/ovn-org/ovn-kubernetes/issues/3506
I would really appreciate if we can hold this PR till IC changes go in - specially if this PR is not important/urgent, conflicts are hard to fix there because we are moving around a lot of code. I don't want to break anything.

File: go-controller/pkg/ovn/egressip.go
Comment (pperiyasamy) [reply go-controller/pkg/ovn/egressip.go]: done

**Comment by pperiyasamy** [comment]
> Do you think we can add a unit test to exercise `syncEgressIPsStatus()` ?

added a test now, hope that covers testing `syncEgressIPsStatus` method.

**Comment by pperiyasamy** [comment]
> flake because of #3506 I would really appreciate if we can hold this PR till IC changes go in - specially if this PR is not important/urgent, conflicts are hard to fix there because we are moving around a lot of code. I don't want to break anything.

Thanks @tssurya for the info, Will wait until #3506 get merged.

**Comment by tssurya** [comment]
> > flake because of #3506 I would really appreciate if we can hold this PR till IC changes go in - specially if this PR is not important/urgent, conflicts are hard to fix there because we are moving around a lot of code. I don't want to break anything.
> 
> Thanks @tssurya for the info, Will wait until #3506 get merged.

also I meant we should get https://github.com/ovn-org/ovn-kubernetes/pull/3386 merged first btw -> the IC changes for egressIP. Because any new tests we add and new things we do, we should do it with the `interconnect=true` flag, else it will be a bit useless for this feature to test on non-IC in some sense..

File: go-controller/pkg/ovn/egressip.go
Comment (tssurya) [reply go-controller/pkg/ovn/egressip.go:1662]: yeah we should wait for IC to land. We are moving the cloud private IP config watcher to the cluster-manager now. So we should add a new sync function for this in cluster-manager. Master will not be watching private IPs anymore.

**Comment by flavio-fernandes** [comment]
/lgtm
/retest

@pperiyasamy if you want, try pulling in pr https://github.com/ovn-org/ovn-kubernetes/pull/3610 for the flake, or keep trying /retest

@tssurya  https://github.com/ovn-org/ovn-kubernetes/pull/3386 is merged; is this okay with you for merging?

File: go-controller/pkg/clustermanager/egressip_controller.go
Comment (tssurya) [inline go-controller/pkg/clustermanager/egressip_controller.go:1059]: `ipNotfound` varable is too generic -> its not conveying the logic. To me it means "egressIP not found in cloudPrivateIPConfigMap cache" but reality is "egressIP not found in cloudPrivateIPConfigMap cache and the status nodeName found on EIP is not the same as we have on cloudPrivateIPConfig" :) -> you see my point? I think this variable naming has to be changed to reflect what it stands for?

File: go-controller/pkg/clustermanager/egressip_controller.go
Comment (tssurya) [inline go-controller/pkg/clustermanager/egressip_controller.go:1071]: nice comments, thanks

File: go-controller/pkg/clustermanager/egressip_controller.go
Comment (tssurya) [inline go-controller/pkg/clustermanager/egressip_controller.go:1073]: the watchers for egressIPs have not yet started at this stage right? so when we start getting the events for adds, its going to be this updated list right?

File: go-controller/pkg/clustermanager/egressip_controller_test.go
Comment (tssurya) [inline go-controller/pkg/clustermanager/egressip_controller_test.go:2194]: ++ THANK YOU for the tests, we are really missing coverage for UTs around non BM platform types..

File: go-controller/pkg/clustermanager/egressip_controller_test.go
Comment (tssurya) [inline go-controller/pkg/clustermanager/egressip_controller_test.go:2210]: nit: would be nice to see more than one in this list so that your loop through statuses logic is checked and one of them remains but other gets deleted -> so that both logic is checked i.e "if its found in cache" and "if its not found in cache",

File: go-controller/pkg/clustermanager/egressip_controller_test.go
Comment (tssurya) [inline go-controller/pkg/clustermanager/egressip_controller_test.go:2229]: +1

File: go-controller/pkg/kube/kube.go
Comment (tssurya) [inline go-controller/pkg/kube/kube.go:379]: we need to fetch from kapi because informers for cloudPrivateIP have not started? hope this list call doesn't cost us a performance at startup

**Review by tssurya — CHANGES_REQUESTED** [review]
We are a bit broken with CloudPrivateIP, we need the second commit here: https://github.com/ovn-org/ovn-kubernetes/pull/3632

File: go-controller/pkg/clustermanager/egressip_controller.go
Comment (pperiyasamy) [reply go-controller/pkg/clustermanager/egressip_controller.go:1087]: done.

File: go-controller/pkg/kube/kube.go
Comment (pperiyasamy) [reply go-controller/pkg/kube/kube.go:379]: yes, i guess this retrieval is done via kube api, but it happens for every egress ip event. this is what more concerning.

File: go-controller/pkg/clustermanager/egressip_controller.go
Comment (pperiyasamy) [reply go-controller/pkg/clustermanager/egressip_controller.go]: done.

**Comment by pperiyasamy** [comment]
> > This commit ensures egress IP status is in sync with cloud private IP config on a restarted ovnk master because stale entry is noticed in the egress IP status though cloud private IP config entry is not present for the given egress IP. This might have happened while processing cloud private IP config delete event, an error is thrown because of bug fixed in PR: #3371.
> 
> Overall I am ok with a clean up stale logic and sync things on start up. But I want to express my concern over how diligently we are doing "PRs to recover" but we need to also pay attention to doing a proper RCA and understanding exactly what led us to this situation rather than the "assumption/might have". But I won't hold this PR for that, have left a few comments above, otherwise I am lgtm.

@tssurya I agree with your concern. When I investigated logs from customer, the relevant logs were not present for impacted IPs because egressIP object for those IPs was created in a previous release long ago, stale status noticed after 2 upgrades. so this fix is done based on mere code study and with background of issue fixed under PR https://github.com/ovn-org/ovn-kubernetes/pull/3371 (which covers only non cloud scenario). We're also trying to get this fix tested in customer setup to confirm this.

File: go-controller/pkg/clustermanager/egressip_controller.go
Comment (tssurya) [reply go-controller/pkg/clustermanager/egressip_controller.go:1079]: the reason why we had the logic called from namespaces is because that is the first watcher we start, so if we had put it in egressIP watcher it would be too late.. but here I think the only 2 objects you care about are egressIPs and CloudPrivateIPs right? Since you don't care about namespaces and pods in this instance, whichever Watcher is started first in CM for EIP we should call it from there to be safe.

File: go-controller/pkg/kube/kube.go
Comment (tssurya) [reply go-controller/pkg/kube/kube.go:379]: wait not sure I follow... the retrieval is done for every event? I thought this is called only once on startup right? Isn't this whole logic a one time cleanup Stale stuff on master restarts a.k.a start ups only?

Comment (pperiyasamy) [reply go-controller/pkg/kube/kube.go:379]: yes, yes. this is being invoked only once at the watcher startup, so we are safe here.

**Comment by flavio-fernandes** [comment]
@pperiyasamy can you pls rebase this pr to use latest/master? we need to understand if ci failures are fixed and unrelated, or are caused by this change.

**Review by tssurya — COMMENTED** [review]
my comments were addressed
before my final lgtm @pperiyasamy did you run this PR against d/s ? We don't run CNCC upstream and we always go blind with changes around CPIPConfig and break things for d/s merges.

**Comment by pperiyasamy** [comment]
> @pperiyasamy did you run this PR against d/s ? We don't run CNCC upstream and we always go blind with changes around CPIPConfig and break things for d/s merges.

OK @tssurya, I just validated the change with a [PR](https://github.com/openshift/ovn-kubernetes/pull/1750) in downstream in an azure cloud, CI seems to be running fine, also verified the cases like egress ips assign/unassign ops, unlabel the egress ip label from one of the node and deleting cloud private ip config while restarting the master.

**Review by tssurya — APPROVED** [review]
/lgtm

File: go-controller/pkg/clustermanager/egressip_controller.go
Comment (jcaamano) [inline go-controller/pkg/clustermanager/egressip_controller.go:1064]: Why are we getting the egressips here if it is what is being received in this mehtod call?

File: go-controller/pkg/clustermanager/egressip_controller.go
Comment (jcaamano) [inline go-controller/pkg/clustermanager/egressip_controller.go:1056]: Wouldn't it make sense for this to be a sync function of the CloudPrivateIPConfigs instead of the EgressIPs? As an added benefit we would have everything we need in the informer cache.

File: go-controller/pkg/clustermanager/egressip_controller_test.go
Comment (pperiyasamy) [reply go-controller/pkg/clustermanager/egressip_controller_test.go:2246]: sure, added a new test to cover no `CloudPrivateIPConfigs` case.

**Review by jcaamano — APPROVED** [review]

**Pull request merged**
